source/MaterialXGenShader/OCIOColorManagementSystem.cpp

```diff
@@ -0,0 +1,634 @@
+                            newInput->setAttribute("height", std::to_string(uniformItem->height));
+                            newInput->setAttribute("depth", std::to_string(uniformItem->depth));
+                            newInput->setAttribute("channels", std::to_string(uniformItem->channelCount));
+                            newInput->setIsUniform(true);
+                        }
+                    }
+                }
+            }
+        }
+        return impl;
+    }
+    catch (const OCIO::Exception& e)
+    {
+        throw ExceptionShaderGenError("Code generation failed for transform: '" + 
+            transform.sourceSpace + "' to '" + transform.targetSpace + "'. Type: '" + transform.type->getName()  + "' : " + e.what());
+    }
+
+    return nullptr;
+}
+
+const ColorManagementResourceMapPtr OCIOColorManagementSystem::getResource(ResourceType resourceType) const
+{
+    if (_ocioInfo)
+    {
+        return _ocioInfo->resourceMap[(int)resourceType];
+    }
+    return nullptr;
+}
+
+void OCIOColorManagementSystem::clearResources()
+{
+    if (_ocioInfo)
+    {
+        _ocioInfo->resourceMap->reset();
+    }
+}
+
+
+//
+// OCIO source code node method
+//
+
+OCIOSourceCodeNode::OCIOSourceCodeNode()
+    : _cmUniforms(ColorManagementSystem::COLOR_MANAGEMENT_UNIFORMS, EMPTY_STRING)
+{
+}
+
+ShaderNodeImplPtr OCIOSourceCodeNode::create()
+{
+    return std::make_shared<OCIOSourceCodeNode>();
+}
+
+void OCIOSourceCodeNode::initialize(const InterfaceElement& element, GenContext& context)
+{
+    SourceCodeNode::initialize(element, context);
+
+    // Cache a uniform block
+    for (auto nodeInput : element.getInputs())
+    {
+        if (nodeInput->getName() != IN_PIXEL_STRING)
+        {
+            string uniformType = nodeInput->getType();
+            if (uniformType == "floatarray" || uniformType == "intarray")
+            {
+                // This will create a sampler. 
+                // NOTE: The value is never set here even though it is storable as is. 
+                _cmUniforms.add(Type::FILENAME, nodeInput->getName());
+            }
+            else
+            {
+                _cmUniforms.add(TypeDesc::get(nodeInput->getType()), nodeInput->getName(), nodeInput->getValue());
+            }
+        }
+    }
+}
+
+void OCIOSourceCodeNode::createVariables(const ShaderNode& /*node*/, GenContext& /*context*/, Shader& shader) const
+{
+    // Add to CM uniform block
+    if (!_cmUniforms.empty())
+    {
+        ShaderStage& stage = shader.getStage(Stage::PIXEL);
+        stage.createUniformBlock(ColorManagementSystem::COLOR_MANAGEMENT_UNIFORMS);
+        VariableBlock& uniformBlock = stage.getUniformBlock(ColorManagementSystem::COLOR_MANAGEMENT_UNIFORMS);
+
+        for (size_t i = 0; i < _cmUniforms.size(); ++i)
+        {
+            ShaderPort* u = const_cast<ShaderPort*>(_cmUniforms[i]);
+            uniformBlock.add(u->getSelf());
```

Comment:
This is where we add uniforms to the appropriate uniform block. Global shader gen logic will determine how to format and emit these. e.g. for some languages we need to create specific resources blocks of uniforms based on type.